File: currency.rb
# currency.rb
require 'net/http'
require 'json'
require 'time'

API_URL = 'https://api.exchangerate.host/latest?base=USD'
CACHE_TTL = 60
$rates = nil
$cache_time = 0
$history = []

def get_rates
  now = Time.now.to_i
  if $rates && (now - $cache_time) < CACHE_TTL
    return $rates
  end
  uri = URI(API_URL)
  begin
    response = Net::HTTP.get_response(uri)
    if response.is_a?(Net::HTTPSuccess)
      data = JSON.parse(response.body)
      if data['success']
        $rates = data['rates']
        $cache_time = now
        return $rates
      else
        raise "API error"
      end
    else
      raise "HTTP error: #{response.code}"
    end
  rescue => e
    puts "Error fetching rates: #{e.message}"
    nil
  end
end

def convert(amount, from_cur, to_cur)
  rates = get_rates
  return nil unless rates
  unless rates.key?(from_cur)
    puts "Currency '#{from_cur}' not supported."
    return nil
  end
  unless rates.key?(to_cur)
    puts "Currency '#{to_cur}' not supported."
    return nil
  end
  usd_amount = from_cur == 'USD' ? amount : amount / rates[from_cur]
  to_cur == 'USD' ? usd_amount : usd_amount * rates[to_cur]
end

def add_history(amount, from_cur, to_cur, result)
  $history << { amount: amount, from: from_cur, to: to_cur, result: result, time: Time.now.strftime('%H:%M:%S') }
  $history.shift if $history.size > 10
end

def show_history
  if $history.empty?
    puts "No conversions yet."
    return
  end
  puts "\n--- History ---"
  $history.each do |h|
    puts "#{h[:time]}  #{'%.2f' % h[:amount]} #{h[:from]} = #{'%.2f' % h[:result]} #{h[:to]}"
  end
end

def interactive
  puts "=== Currency Converter ==="
  loop do
    puts "\n1. Convert"
    puts "2. Show history"
    puts "3. Exit"
    print "Choose: "
    choice = gets.chomp.strip
    case choice
    when '1'
      print "Amount: "
      amount_str = gets.chomp.strip
      begin
        amount = Float(amount_str)
      rescue ArgumentError
        puts "Invalid amount."
        next
      end
      if amount < 0
        puts "Invalid amount."
        next
      end
      print "From (USD): "
      from_cur = gets.chomp.strip.upcase
      from_cur = 'USD' if from_cur.empty?
      print "To: "
      to_cur = gets.chomp.strip.upcase
      if to_cur.empty?
        puts "Please enter a target currency."
        next
      end
      result = convert(amount, from_cur, to_cur)
      if result
        puts "#{'%.2f' % amount} #{from_cur} = #{'%.2f' % result} #{to_cur}"
        add_history(amount, from_cur, to_cur, result)
      end
    when '2'
      show_history
    when '3'
      puts "Goodbye!"
      break
    else
      puts "Invalid choice."
    end
  end
end

def cli
  if ARGV.size != 3
    puts "Usage: ruby currency.rb <amount> <from_currency> <to_currency>"
    exit 1
  end
  begin
    amount = Float(ARGV[0])
  rescue ArgumentError
    puts "Invalid amount."
    exit 1
  end
  if amount < 0
    puts "Invalid amount."
    exit 1
  end
  from_cur = ARGV[1].upcase
  to_cur = ARGV[2].upcase
  result = convert(amount, from_cur, to_cur)
  if result
    puts "#{'%.2f' % amount} #{from_cur} = #{'%.2f' % result} #{to_cur}"
  end
end

if ARGV.empty?
  interactive
else
  cli
end

File: currency.py
# currency.py
import sys
import json
import time
import urllib.request
import urllib.error
from datetime import datetime

API_URL = "https://api.exchangerate.host/latest?base=USD"
CACHE_TTL = 60  # seconds
_cache = {"rates": None, "timestamp": 0}
_history = []

def get_rates():
    now = time.time()
    if _cache["rates"] is not None and (now - _cache["timestamp"]) < CACHE_TTL:
        return _cache["rates"]
    try:
        with urllib.request.urlopen(API_URL, timeout=5) as response:
            data = json.loads(response.read().decode())
            if data.get("success"):
                _cache["rates"] = data["rates"]
                _cache["timestamp"] = now
                return _cache["rates"]
            else:
                print("API error")
                return None
    except Exception as e:
        print(f"Error fetching rates: {e}")
        return None

def convert(amount, from_cur, to_cur):
    rates = get_rates()
    if rates is None:
        return None
    if from_cur not in rates:
        print(f"Currency '{from_cur}' not supported.")
        return None
    if to_cur not in rates:
        print(f"Currency '{to_cur}' not supported.")
        return None
    usd_amount = amount / rates[from_cur] if from_cur != "USD" else amount
    result = usd_amount * rates[to_cur] if to_cur != "USD" else usd_amount
    return result

def add_history(amount, from_cur, to_cur, result):
    _history.append((amount, from_cur, to_cur, result, datetime.now().strftime("%H:%M:%S")))
    if len(_history) > 10:
        _history.pop(0)

def show_history():
    if not _history:
        print("No conversions yet.")
        return
    print("\n--- History ---")
    for h in _history:
        print(f"{h[4]}  {h[0]:.2f} {h[1]} = {h[3]:.2f} {h[2]}")

def interactive():
    print("=== Currency Converter ===")
    while True:
        print("\n1. Convert")
        print("2. Show history")
        print("3. Exit")
        choice = input("Choose: ").strip()
        if choice == "1":
            try:
                amount = float(input("Amount: "))
                from_cur = input("From (USD): ").strip().upper() or "USD"
                to_cur = input("To: ").strip().upper()
                if not to_cur:
                    print("Please enter a target currency.")
                    continue
                result = convert(amount, from_cur, to_cur)
                if result is not None:
                    print(f"{amount:.2f} {from_cur} = {result:.2f} {to_cur}")
                    add_history(amount, from_cur, to_cur, result)
            except ValueError:
                print("Invalid amount.")
        elif choice == "2":
            show_history()
        elif choice == "3":
            print("Goodbye!")
            break
        else:
            print("Invalid choice.")

def cli():
    if len(sys.argv) != 4:
        print("Usage: python currency.py <amount> <from_currency> <to_currency>")
        sys.exit(1)
    try:
        amount = float(sys.argv[1])
        from_cur = sys.argv[2].upper()
        to_cur = sys.argv[3].upper()
        result = convert(amount, from_cur, to_cur)
        if result is not None:
            print(f"{amount:.2f} {from_cur} = {result:.2f} {to_cur}")
    except ValueError:
        print("Invalid amount.")

if __name__ == "__main__":
    if len(sys.argv) == 1:
        interactive()
    else:
        cli()
